CONTRIBUTING.md

```diff
@@ -14,6 +14,204 @@ Here are some useful hints to get involved.
-* We accept github pull requests and this is the preferred way to contribute to CRIU. If you prefer to send patches by email, you are welcome to send them to [the devel list](http://criu.org/How_to_submit_patches);
+* We accept github pull requests and this is the preferred way to contribute to CRIU. If you prefer to send patches by email, you are welcome to send them to [CRIU development mailing list](https://lists.openvz.org/mailman/listinfo/criu).
+Below we describe in more detail recommend practices for CRIU developemnt.
 * Spread the word about CRIU in [social networks](http://criu.org/Contacts);
 * If you're giving a talk about CRIU -- let us know, we'll mention it on the [wiki main page](https://criu.org/News/events);
+
+### Seting up the developemnt environment
+
+Although criu could be run as non-root (see [Security](https://criu.org/Security), development is better to be done as root. For example, some tests require root. So, it would be a good idea to set up some recent Linux distro on a virtual machine.
+
+### Get the source code
+
+The CRIU sources are tracked by Git. Official CRIU repo is at https://github.com/checkpoint-restore/criu.
+
+The repository may contain multiple branches. Development happens in the **criu-dev** branch.
+
+To clone CRIU repo and switch to the proper branch, run:
+
+```
+        git clone https://github.com/checkpoint-restore/criu criu
+        cd criu
+        git checkout criu-dev
+```
+
+### Compile
+
+First, you need to install compile-time dependencies. Check [Installation dependencies](https://criu.org/Installation#Dependencies) for more info.
+
+To compile CRIU, run:
+
+```
+        make
+```
+
+This should create the `./criu/criu` executable.
+
+## Edit the source code
+
+If you use ctags, you can generate the ctags file by running
+
+```
+        make tags
+```
+
+When you change the source code, please keep in mind the following code conventions:
+
+* we prefer tabs and indentations to be 8 characters width
+* CRIU mostly follows [Linux kernel coding style](https://www.kernel.org/doc/Documentation/process/coding-style.rst), but we are less strict than the kernel community.
+
+Other conventions can be learned from the source code itself. In short, make sure your new code
+looks similar to what is already there.
+
+## Test your changes
+
+CRIU comes with an extensive test suite. To check whether your changes introduce any regressions, run
+
+```
+         make test
+```
+
+The command runs [ZDTM Test Suite](https://criu.org/ZDTM_Test_Suite). Check for any error messages produced by it.
+
+In case you'd rather have someone else run the tests, you can use travis-ci for your
+own github fork of CRIU. It will check the compilation for various supported platforms,
+as well as run most of the tests from the suite. See https://travis-ci.org/checkpoint-restore/criu
+for more details.
+
+## Sign your work
```

Comment:
we need to add a section which explains how to split changes on patches and how to write commit messages.

Reply:
I would recommend this article https://chris.beams.io/posts/git-commit/